### DISCUSSION TITLE
Question how to bulk patch

### DISCUSSION THREAD

**Discussion by coolfox89** [emmercm/igir#1350] [Q&A]
### Is your feature request related to a problem?

I need help with bulk patching. I’m a beginner with scripts and code so I’m not sure how to apply this.

### Describe the solution you'd like

I would like to bulk patch the Romhacking.net archive since its recent shutdown. How do I use your tool to bulk patch? Can you send me an example of what a bulk patch would look like?

The Romhacking.net archive has a SQL file that contains hash values for the patches. Would your tool be able to work with the SQL file to help automate the patching process?

**Comment by emmercm** [comment]
I haven't looked through the dump yet myself, but this is interesting.

Right now your options are:

1. For patch types that contain a CRC32 in their file contents, those should just work.
2. For patch types that _don't_ contain a CRC32 in their file contents (which I expect to be most of them since IPS seems to be most common), you're going to need some kind of script.

It looks like the DB dump is MySQL, so a person would need to do something like:

1. Start up a MySQL instance, load the backup file
2. Query the various tables for: hack name, original ROM CRC32, and patch filename
3. Extract each of the archived patches, and rename the patch file to `<hack name> <original ROM CRC32>.<patch extension>`

I don't personally have the time to handle this, but I do think someone can do it with some work.

**Reply by coolfox89** [reply]
I would like to try taking this project on. What script do I run to automate the patching? I want to test this out on a few patches to see if it will work. Do I need to type something in the command line?

**Reply by Europia79** [reply]
This is ill-advised: ...I have personal, first-hand experience patching thousands upon thousands of ROMs and I can tell you that there are just way too many errors on RHDN that you'll just end up with another "GoodSet":  ...And THAT was one of the primary complaints with the GoodSets:  ...It had bad romhacks where patches were applied incorrectly & pushed to "production" without ever testing them to confirm that they were actually working.

Altho, I do remember reading an article a long time ago by an NES Developer who had an idea to automate the testing process of romhacks:  ...Unfortunately, I do not remember the Author or Source...  But if you can find it, then that will help with automating your script.

Alternatively, you could manually patch & test them.  It would be grueling & time-consuming, but once done, you could setup bulk patching for everyone else to benefit from.

But now that I think about it... another alternative would be to output a "List of ROMs to Test" after your patching script runs:  ...Where you have the script use that List to either directly launch the Emulator (if there's a CLI), or, you have it open the directory location where the file resides:  ...Then, if you have file associations setup, then you can double-click that file to launch it.

Then, after "testing" that it works, you close the emulator & go back to the script & tell it if the romhack was "good" or bad:  ...If "good", then your script removes it from the List of Tests.  If "bad", then your script removes it from your Testing List and adds it (and the corresponding patch) to an "Investigation Queue": ...Where more inspection is required to see what the problems were.
